crates/sqlbuiltins/src/functions/scalars/openai.rs

```diff
@@ -0,0 +1,245 @@
+                let scalar = creds_from_arg(DEFAULT_CREDENTIAL_LOCATION.clone());
+                (scalar, model, 0)
+            }
+            // openai_embed(<model>, <expr>)
+            2 => {
+                let model = model_from_arg(&args[0])?;
+                let scalar = creds_from_arg(DEFAULT_CREDENTIAL_LOCATION.clone());
+                (scalar, model, 1)
+            }
+            // openai_embed('api_key', '<model>', '<expr>')
+            3 => {
+                let creds = match args.get(0) {
+                    Some(Expr::Literal(ScalarValue::Utf8(v))) => {
+                        Some(v.clone().unwrap().to_string())
+                    }
+                    Some(Expr::ScalarVariable(_, values)) => {
+                        let scalar = creds_from_arg(values.clone());
+                        scalar
+                    }
+                    _ => return Err(DataFusionError::Plan("Invalid argument".to_string())),
+                };
+
+                let model = model_from_arg(&args[1])?;
+                (creds, model, 2)
+            }
+            _ => return Err(DataFusionError::Plan("Invalid argument count".to_string())),
+        };
+        let Some(creds) = creds else {
+            return Err(DataFusionError::Plan(
+                "No API key or credential provided".to_string(),
+            ));
+        };
+
+        let model_len = model.len();
+        let return_type_fn: ReturnTypeFunction = Arc::new(move |_| {
+            let f = Field::new("item", DataType::Float32, true);
+            let dtype = DataType::FixedSizeList(Arc::new(f), model_len);
+            Ok(Arc::new(dtype))
+        });
+
+        let scalar_fn_impl: ScalarFunctionImplementation = Arc::new(move |args| {
+            let input = match &args[0] {
+                ColumnarValue::Array(arr) => match arr.data_type() {
+                    DataType::Utf8 => EmbeddingInput::StringArray(
+                        arr.as_string::<i32>()
+                            .into_iter()
+                            .map(|s| s.unwrap_or_default().to_string())
+                            .collect(),
+                    ),
+                    DataType::LargeUtf8 => EmbeddingInput::StringArray(
+                        arr.as_string::<i64>()
+                            .into_iter()
+                            .map(|s| s.unwrap_or_default().to_string())
+                            .collect(),
+                    ),
+                    _ => return Err(DataFusionError::Plan("Invalid argument".to_string())),
+                },
+                ColumnarValue::Scalar(ScalarValue::Utf8(v) | ScalarValue::LargeUtf8(v)) => {
+                    EmbeddingInput::StringArray(vec![v.clone().unwrap_or_default()])
+                }
+                _ => return Err(DataFusionError::Plan("Invalid argument".to_string())),
+            };
+
+            let client = Client::with_config(OpenAIConfig::new().with_api_key(creds.clone()));
```

Comment:
is there a way to specify a timeout?

Reply:
probably, this just uses reqwest clients under the hood, so timeouts should be doable. 

We can add timeouts in a followup.

Reply:
https://docs.rs/async-openai/latest/async_openai/struct.Client.html#method.with_http_client <- yeah seems straightforward.